dialogue-core/src/test/java/com/palantir/dialogue/core/StickyEndpointChannels2Test.java

```diff
@@ -0,0 +1,160 @@
+                .isNotEqualTo(QueueAttachments.getQueueOverride(request2));
+    }
+
+    @Test
+    public void requests_queue_up_and_propagate_sticky_target() throws ExecutionException {
+        Channel channel = sticky.get();
+
+        Request request1 = Request.builder().build();
+        SettableFuture<Response> response1SettableFuture = expectAddStickyTokenRequest(request1);
+        ListenableFuture<Response> response1ListenableFuture = channel.execute(endpoint, request1);
+        assertThat(response1ListenableFuture).isNotDone();
+
+        Request request2 = Request.builder().build();
+        SettableFuture<Response> response2SettableFuture = expectStickyRequest(response1ListenableFuture, request2);
+        ListenableFuture<Response> response2ListenableFuture = channel.execute(endpoint, request2);
+        assertThat(response2ListenableFuture).isNotDone();
+
+        TestResponse testResponse1 = TestResponse.withBody(null);
+        response1SettableFuture.set(testResponse1);
+
+        assertThat(Futures.getDone(response1ListenableFuture)).isEqualTo(testResponse1);
+
+        TestResponse testResponse2 = TestResponse.withBody(null);
+        response2SettableFuture.set(testResponse2);
+        assertThat(Futures.getDone(response2ListenableFuture)).isEqualTo(testResponse2);
+    }
+
+    // On failure we execute queued request.
+    // We queue up requests.
+
+    private SettableFuture<Response> expectAddStickyTokenRequest(Request request) {
+        SettableFuture<Response> responseSettableFuture = SettableFuture.create();
+        when(endpointChannel.execute(request)).thenAnswer((Answer<ListenableFuture<Response>>) invocation -> {
+            Request actualRequest = invocation.getArgument(0);
+            assertThat(actualRequest).isEqualTo(request);
+            LimitedChannel stickyTarget = mock(LimitedChannel.class);
+            when(stickyTarget.maybeExecute(endpoint, request, LimitEnforcement.DEFAULT_ENABLED))
+                    .thenReturn(Optional.of(responseSettableFuture));
+            return StickyAttachments.maybeAddStickyToken(
+                            stickyTarget, endpoint, actualRequest, LimitEnforcement.DEFAULT_ENABLED)
+                    .get();
+        });
+        return responseSettableFuture;
+    }
+
+    private SettableFuture<Response> expectStickyRequest(ListenableFuture<Response> response, Request request) {
+        SettableFuture<Response> responseSettableFuture = SettableFuture.create();
+        when(endpointChannel.execute(request)).thenAnswer((Answer<ListenableFuture<Response>>) invocation -> {
+            Request actualRequest = invocation.getArgument(0);
+            assertThat(actualRequest).isEqualTo(request);
+            assertThat(response).isDone();
+            assertThat(Futures.getDone(response).attachments().getOrDefault(StickyAttachments.STICKY_TOKEN, null))
+                    .isEqualTo(request.attachments().getOrDefault(StickyAttachments.STICKY, null));
```

Comment:
We could add a sticky-integrity-validating-channel to the default chain -- when no sticky attachment is on the request, we forward the request as usual, but if there's a sticky token, we assert the sticky response attachment appears on the response. Same for the initial sticky request -> sticky_token. That way if a configuration we haven't thought to test occurs, we can fail fast rather than producing incorrect behavior.

Reply:
Ack

Reply:
Done